### PR TITLE
removing map marhsals from core

### PR DESCRIPTION
**Pull request by akshaydeo**
## Summary

Replace `map[string]interface{}` and `interface{}` types with `json.RawMessage` for JSON data handling across provider implementations to preserve key ordering and improve prompt caching effectiveness.

## Changes

- Changed tool input/output fields from `map[string]interface{}` to `json.RawMessage` in Anthropic, Bedrock, Cohere, Gemini, and Replicate providers
- Updated JSON parsing logic to use `json.Unmarshal` with `json.RawMessage` instead of type assertions
- Modified tool argument handling to preserve original JSON structure without re-marshaling through Go maps
- Added JSON validation using `json.Valid()` before processing tool arguments
- Updated structured output format fields to use `json.RawMessage` for consistent key ordering

## Type of change

- [x] Refactor
- [ ] Bug fix
- [ ] Feature
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [ ] Transports (HTTP)
- [x] Providers/Integrations
- [ ] Plugins
- [ ] UI (Next.js)
- [ ] Docs

## How to test

Validate that tool calls and structured outputs maintain consistent JSON formatting:

```sh
# Core/Transports
go version
go test ./...

# Test tool calling functionality
go test ./core/providers/anthropic/... -v
go test ./core/providers/bedrock/... -v
go test ./core/providers/gemini/... -v

# Verify JSON key ordering preservation in tool inputs
go test -run TestToolInputPreservation ./core/providers/...
```

Test with tool-calling requests to ensure arguments are properly preserved and prompt caching works effectively.

## Screenshots/Recordings

N/A

## Breaking changes

- [ ] Yes
- [x] No

This is an internal refactor that maintains API compatibility while improving JSON handling consistency.

## Related issues

N/A

## Security considerations

No security implications. This change improves data consistency and reduces potential JSON parsing edge cases.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [x] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [x] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by akshaydeo** [comment]
* **#2184** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2182** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2174** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2174?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2174?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: core/providers/anthropic/responses.go [hunk 1]
```diff
@@ -3077,8 +3077,11 @@ func ConvertBifrostMessagesToAnthropicMessages(ctx *schemas.BifrostContext, bifr
 			}
 			if msg.ResponsesToolMessage != nil && msg.ResponsesToolMessage.Action != nil &&
 				msg.ResponsesToolMessage.Action.ResponsesWebFetchToolCallAction != nil {
-				serverToolUseBlock.Input = map[string]interface{}{
+				inputBytes, err := json.Marshal(map[string]interface{}{
 					"url": msg.ResponsesToolMessage.Action.ResponsesWebFetchToolCallAction.URL,
+				})
+				if err == nil {
+					serverToolUseBlock.Input = json.RawMessage(inputBytes)
 				}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/anthropic/responses.go]: Fixed — replaced all `json.Marshal` and `sonic.Marshal` calls with `providerUtils.MarshalSorted` (uses `sonic.Config{SortMapKeys: true}`) across all PR-changed files. Also replaced `json.Unmarshal` with `sonic.Unmarshal`, and converted simple field extractions from map unmarshals to `gjson` via `providerUtils.GetJSONField` for zero-allocation reads. See updated code.

File: core/providers/bedrock/utils.go [hunk 2]
```diff
@@ -862,12 +862,13 @@ func convertResponseFormatToTool(ctx *schemas.BifrostContext, params *schemas.Ch
...
 
 	// Create the Bedrock tool
+	schemaObjBytes, _ := json.Marshal(schemaObj)
 	return &BedrockTool{
 		ToolSpec: &BedrockToolSpec{
 			Name:        toolName,
 			Description: schemas.Ptr(description),
 			InputSchema: BedrockToolInputSchema{
-				JSON: schemaObj,
+				JSON: json.RawMessage(schemaObjBytes),
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/utils.go]: Fixed — all `json.Marshal` calls in bedrock/utils.go replaced with `providerUtils.MarshalSorted`. See updated code.

File: core/providers/bedrock/utils.go [hunk 3]
```diff
@@ -1468,13 +1472,16 @@ func tryParseJSONIntoContentBlock(text string) BedrockContentBlock {
...
 		case []any:
 			// Arrays need to be wrapped
-			return BedrockContentBlock{JSON: map[string]any{"results": v}}
+			jsonBytes, _ := json.Marshal(map[string]any{"results": v})
+			return BedrockContentBlock{JSON: json.RawMessage(jsonBytes)}
 		default:
 			// Primitives (string, number, boolean, null) need to be wrapped
-			return BedrockContentBlock{JSON: map[string]any{"value": v}}
+			jsonBytes, _ := json.Marshal(map[string]any{"value": v})
+			return BedrockContentBlock{JSON: json.RawMessage(jsonBytes)}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/utils.go]: Fixed — both `tryParseJSONIntoContentBlock` and the inline equivalent at line ~510 now use `json.Compact` to validate and preserve original bytes. Objects keep their key order, arrays/primitives get wrapped with raw byte concatenation — zero map allocations. See updated code.

File: core/providers/gemini/videos.go [hunk 4]
```diff
@@ -232,15 +233,18 @@ func ToBifrostVideoGenerationResponse(operation *GenerateVideosOperation, model
...
+				}
 			}
 		}
 	} else if operation.Error != nil {
 		response.Status = schemas.VideoStatusFailed
 		response.Error = &schemas.VideoCreateError{
 			Code:    "video_generation_failed",
-			Message: fmt.Sprintf("%v", operation.Error),
+			Message: string(operation.Error),
 		}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/gemini/videos.go:254]: Fixed — now using `providerUtils.GetJSONField` (gjson) to extract structured `code` and `message` fields from `operation.Error` instead of using the raw JSON blob as the message. Falls back to raw string if fields are missing. See updated code.

File: core/providers/bedrock/utils.go [hunk 5]
```diff
@@ -518,18 +518,21 @@ func convertToolMessages(msgs []schemas.ChatMessage) (BedrockMessage, error) {
...
+						JSON: json.RawMessage(jsonBytes),
 					})
 				default:
 					// Primitives (string, number, boolean, null) need to be wrapped
+					jsonBytes, _ := json.Marshal(map[string]any{"value": v})
 					toolResultContent = append(toolResultContent, BedrockContentBlock{
-						JSON: map[string]any{"value": v},
+						JSON: json.RawMessage(jsonBytes),
 					})
 				}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/utils.go]: Fixed — this was addressed together with comment #3. Both locations now use `json.Compact` + raw byte wrapping instead of parse-remarshal. See updated code.

File: core/providers/cohere/rerank.go [hunk 6]
```diff
@@ -92,25 +93,28 @@ func (response *CohereRerankResponse) ToBifrostRerankResponse(documents []schema
...
+				for k, v := range docMap {
+					if k != "text" && k != "id" {
+						meta[k] = v
+					}
+				}
+				if len(meta) > 0 {
+					doc.Meta = meta
+				}
+				rerankResult.Document = doc
 			}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/cohere/rerank.go:132]: Fixed — the parser now unwraps `"metadata"` and `"meta"` keys before assigning to `doc.Meta`, preventing the nested wrapper issue on round-trip. Other top-level keys are still merged in. See updated code.

File: core/providers/cohere/responses.go [hunk 7]
```diff
@@ -1,6 +1,7 @@
 package cohere
 
 import (
+	"encoding/json"
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/cohere/responses.go]: Fixed — replaced `json.Unmarshal` + `map[string]interface{}` with `providerUtils.GetJSONField` (gjson) for zero-allocation field extraction. Also removed the now-unused `encoding/json` import. See updated code.

File: core/providers/gemini/responses.go [hunk 8]
```diff
@@ -348,9 +348,10 @@ func ToGeminiResponsesResponse(bifrostResp *schemas.BifrostResponsesResponse) *G
 					funcName = *msg.ResponsesToolMessage.CallID
 				}
 
+				responseBytes, _ := sonic.Marshal(responseMap)
 				functionResponse := &FunctionResponse{
 					Name:     funcName,
-					Response: responseMap,
+					Response: json.RawMessage(responseBytes),
 				}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/gemini/responses.go]: Fixed — both locations now check `json.Valid()` and use `json.RawMessage` for valid JSON outputs, preserving structure without double-quoting. Plain strings fall through as before. See updated code.

File: core/providers/gemini/responses.go [hunk 9]
```diff
@@ -1917,13 +1918,18 @@ func convertGeminiContentsToResponsesMessages(contents []Content) []schemas.Resp
...
+					var respMap map[string]any
+					if err := sonic.Unmarshal([]byte(part.FunctionResponse.Response), &respMap); err == nil {
+						if output, ok := respMap["output"].(string); ok {
+							responseStr = output
+						} else {
+							responseStr = string(part.FunctionResponse.Response)
+						}
+					} else {
+						responseStr = string(part.FunctionResponse.Response)
 					}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/gemini/responses.go]: Fixed — replaced the manual `sonic.Unmarshal` + map extraction with `providerUtils.GetJSONField` (gjson). This correctly handles both string and object outputs, and removed the now-unused `sonic` import from responses.go. See updated code.

File: core/providers/gemini/utils.go [hunk 10]
```diff
@@ -1551,16 +1551,19 @@ func convertBifrostMessagesToGemini(messages []schemas.ChatMessage) ([]Content,
...
+				var tmp any
+				if err := sonic.Unmarshal([]byte(contentStr), &tmp); err == nil {
+					responseData = json.RawMessage(contentStr)
+				} else {
 					// If unmarshaling fails, wrap the original string to preserve it
-					responseData = map[string]any{
+					responseData, _ = sonic.Marshal(map[string]any{
 						"content": contentStr,
-					}
+					})
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/gemini/utils.go]: Fixed — replaced `sonic.Unmarshal` validation with `json.Compact` + first-byte check. Only JSON objects (`{...}`) go through the raw path; arrays, scalars, and invalid JSON get wrapped in `{"content": ...}` to comply with Gemini's Struct requirement. See updated code.

File: core/providers/anthropic/responses.go [hunk 11]
```diff
@@ -968,12 +969,9 @@ func (chunk *AnthropicStreamEvent) ToBifrostResponsesStream(ctx context.Context,
...
-					var inputMap map[string]interface{}
-					if err := json.Unmarshal([]byte(state.AccumulatedJSON), &inputMap); err == nil {
-						if q, ok := inputMap["query"].(string); ok {
-							query = q
-							queries = []string{q}
-						}
+					if q := providerUtils.GetJSONField([]byte(state.AccumulatedJSON), "query"); q.Exists() {
+						query = q.String()
+						queries = []string{q.String()}
 					}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/anthropic/responses.go]: Fixed - Added `q.Type == gjson.String` / `u.Type == gjson.String` type checks and switched from `.String()` to `.Str` at all 5 locations (lines 973, 3476, 3489, 3776, 3800). This prevents type coercion from non-string JSON values. See updated code.

File: core/providers/anthropic/responses.go [hunk 12]
```diff
@@ -3462,39 +3463,37 @@ func convertAnthropicContentBlocksToResponsesMessagesGrouped(contentBlocks []Ant
...
+				if u := providerUtils.GetJSONField(toolBlock.Input, "url"); u.Exists() {
+					bifrostMsg.ResponsesToolMessage.Action = &schemas.ResponsesToolMessageActionStruct{
+						ResponsesWebFetchToolCallAction: &schemas.ResponsesWebFetchToolCallAction{
+							URL: u.String(),
+						},
 					}
 				}
 			} else {
-				bifrostMsg.ResponsesToolMessage.Arguments = schemas.Ptr(schemas.JsonifyInput(toolBlock.Input))
+				bifrostMsg.ResponsesToolMessage.Arguments = schemas.Ptr(string(toolBlock.Input))
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/anthropic/responses.go]: Fixed - Added `len() > 0` guards before `string()` conversion at all 3 locations (generic tool use, computer tool else branch, MCP tool use). This preserves nil vs empty-string semantics for the Arguments field. See updated code.

File: core/providers/bedrock/responses.go [hunk 13]
```diff
@@ -1419,12 +1420,18 @@ func (request *BedrockConverseRequest) ToBifrostResponsesRequest(ctx *schemas.Bi
...
+					if err := sonic.Unmarshal(tool.ToolSpec.InputSchema.JSON, &params); err == nil && params.Type != "" {
+						bifrostTool.ResponsesToolFunction.Parameters = &params
+					} else {
+						// Fallback: unmarshal as map and convert
+						var paramsMap map[string]interface{}
+						if err := sonic.Unmarshal(tool.ToolSpec.InputSchema.JSON, &paramsMap); err == nil {
+							params := convertMapToToolFunctionParameters(paramsMap)
+							bifrostTool.ResponsesToolFunction.Parameters = params
+						}
+					}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/responses.go:1434]: Fixed - Removed the `params.Type != ""` check. Now any successful `sonic.Unmarshal` into `ToolFunctionParameters` is trusted, which correctly handles schemas using `$ref`, `anyOf`, `oneOf`, or `allOf` without a top-level `type` field. See updated code.

File: core/providers/bedrock/utils.go [hunk 14]
```diff
@@ -862,12 +868,13 @@ func convertResponseFormatToTool(ctx *schemas.BifrostContext, params *schemas.Ch
...
 
 	// Create the Bedrock tool
+	schemaObjBytes, _ := providerUtils.MarshalSorted(schemaObj)
 	return &BedrockTool{
 		ToolSpec: &BedrockToolSpec{
 			Name:        toolName,
 			Description: schemas.Ptr(description),
 			InputSchema: BedrockToolInputSchema{
-				JSON: schemaObj,
+				JSON: json.RawMessage(schemaObjBytes),
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/utils.go]: Fixed - Added error checks on all 3 `MarshalSorted` calls. The two tool-conversion functions (`convertResponseFormatToTool`, `convertTextFormatToTool`) now return `nil` on failure, and the `convertToolConfig` loop now `continue`s past the failing tool. This prevents empty `inputSchema:{}` from reaching Bedrock. See updated code.

File: core/providers/bedrock/bedrock_test.go [hunk 15]
```diff
@@ -14,6 +14,36 @@ import (
...
+	if err := json.Unmarshal(expected, &e); err != nil {
+		t.Errorf("failed to unmarshal expected JSON: %v", err)
+		return
+	}
+	if err := json.Unmarshal(actual, &a); err != nil {
+		t.Errorf("failed to unmarshal actual JSON: %v", err)
+		return
+	}
+	assert.Equal(t, e, a, msgAndArgs...)
+}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/bedrock_test.go:38]: This is a valid improvement but out of scope for this PR. The helpers (`mustMarshalJSON`, `jsonEqual`, `mustMarshalToolParams`) are already in place — adding comprehensive structural assertions for ToolUse.Input across all response-conversion cases will be done as a focused follow-up.

File: core/providers/bedrock/responses.go [hunk 16]
```diff
@@ -1899,12 +1906,13 @@ func ToBedrockResponsesRequest(ctx *schemas.BifrostContext, bifrostReq *schemas.
...
 				}
 
+				schemaObjectBytes, _ := providerUtils.MarshalSorted(schemaObject)
 				bedrockTool := BedrockTool{
 					ToolSpec: &BedrockToolSpec{
 						Name:        name,
 						Description: &description,
 						InputSchema: BedrockToolInputSchema{
-							JSON: schemaObject,
+							JSON: json.RawMessage(schemaObjectBytes),
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/responses.go]: Fixed - Added error check on `MarshalSorted` and now returning a descriptive error: `fmt.Errorf("failed to serialize tool schema %q: %w", name, err)`. This prevents silently emitting invalid `inputSchema` on the Bedrock request. See updated code.

File: core/providers/bedrock/responses.go [hunk 17]
```diff
@@ -2980,7 +2989,7 @@ func convertSingleBedrockMessageToBifrostMessages(ctx *schemas.BifrostContext, m
 				// Marshal the tool input to JSON string
 				var contentStr string
 				if block.ToolUse.Input != nil {
-					contentStr = schemas.JsonifyInput(block.ToolUse.Input)
+					contentStr = string(block.ToolUse.Input)
 				} else {
 					contentStr = "{}"
 				}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/responses.go:2998]: Fixed - Normalized nil tool input to `"{}"` in the normal tool-call branch, matching the structured-output branch behavior. When `block.ToolUse.Input` is nil, `Arguments` now gets `"{}"` instead of `""`. See updated code.

File: core/providers/anthropic/types.go [hunk 18]
```diff
@@ -71,7 +72,7 @@ func (req *AnthropicTextRequest) IsStreamingRequested() bool {
 // AnthropicOutputConfig represents the GA structured outputs config (output_config.format)
 // and the effort parameter (output_config.effort) for controlling token spending.
 type AnthropicOutputConfig struct {
-	Format interface{} `json:"format,omitempty"`
+	Format json.RawMessage `json:"format,omitempty"`
 	Effort *string     `json:"effort,omitempty"` // "low", "medium", "high", "max" (Opus 4.5+)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/anthropic/types.go:76]: Already addressed — `OutputFormat` and `OutputConfig.Format` are compacted in `UnmarshalJSON` (lines 421–433). `AnthropicToolInputExample.Input` is already `json.RawMessage`, not `any`. This is solved, can you check and resolve if done properly?

File: core/providers/bedrock/batch.go [hunk 19]
```diff
@@ -165,9 +166,12 @@ func parseBatchResultsJSONL(content []byte, provider *BedrockProvider) ([]schema
...
-			resultItem.Response = &schemas.BatchResultResponse{
-				StatusCode: 200,
-				Body:       bedrockResult.ModelOutput,
+			var bodyMap map[string]interface{}
+			if err := json.Unmarshal(bedrockResult.ModelOutput, &bodyMap); err == nil {
+				resultItem.Response = &schemas.BatchResultResponse{
+					StatusCode: 200,
+					Body:       bodyMap,
+				}
 			}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/batch.go:180]: Already addressed — the `else` block (lines 176–180) catches decode failures and records them as `resultItem.Error` with a `parse_error` code. Returning the error directly would abort the entire batch on a single record failure, which is worse. The current per-item error recording is intentional. This is solved, can you check and resolve if done properly?

File: core/providers/bedrock/bedrock_test.go [hunk 20]
```diff
@@ -14,6 +14,36 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func mustMarshalJSON(v interface{}) json.RawMessage {
+	b, _ := json.Marshal(v)
+	return json.RawMessage(b)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/bedrock_test.go:19]: Valid point — will fix in a follow-up. These helpers only marshal static test fixtures so the risk is near-zero, but aligning with the `must*` panic convention is the right call.

File: core/providers/cohere/rerank.go [hunk 21]
```diff
@@ -92,25 +93,37 @@ func (response *CohereRerankResponse) ToBifrostRerankResponse(documents []schema
...
+				}
+				for k, v := range docMap {
+					if k != "text" && k != "id" && k != "metadata" && k != "meta" {
+						meta[k] = v
+					}
+				}
+				if len(meta) > 0 {
+					doc.Meta = meta
+				}
+				rerankResult.Document = doc
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/cohere/rerank.go]: Fixed — added a `populated` flag to only attach `rerankResult.Document` when at least one field (text, id, or metadata) was actually extracted. Empty payloads like `{}` no longer produce ghost documents. This is solved, can you check and resolve if done properly?

File: core/providers/cohere/responses.go [hunk 22]
```diff
@@ -878,16 +878,21 @@ func (chunk *CohereStreamEvent) ToBifrostResponsesStream(sequenceNumber int, sta
...
+						annotation.FileID = &idStr
 					}
-					if snippet, ok := (*source.Document)["snippet"].(string); ok {
+					if s := providerUtils.GetJSONField(doc, "snippet"); s.Exists() {
+						snippet := s.String()
 						annotation.Text = &snippet
 					}
-					if url, ok := (*source.Document)["url"].(string); ok {
+					if u := providerUtils.GetJSONField(doc, "url"); u.Exists() {
+						url := u.String()
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/cohere/responses.go:896]: Fixed — added `t.Type == gjson.String` checks on all four citation field extractions (title, id, snippet, url) to prevent coercion of non-string JSON values. This is solved, can you check and resolve if done properly?

File: core/providers/gemini/utils.go [hunk 23]
```diff
@@ -2323,18 +2326,17 @@ func extractFunctionResponseOutput(funcResp *FunctionResponse) string {
...
 	}
 
-	// If no "output" key, marshal the entire response
-	if jsonResponse, err := sonic.Marshal(funcResp.Response); err == nil {
-		return string(jsonResponse)
-	}
-
-	return ""
+	// If no "output" key or unmarshal failed, return raw JSON
+	return string(funcResp.Response)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/gemini/utils.go:2341]: Fixed — changed to `map[string]json.RawMessage` so non-string `output` payloads (objects, arrays) are correctly unwrapped from the envelope instead of returning the whole response. This is solved, can you check and resolve if done properly?

**Comment by akshaydeo** [comment]
@coderabbitai full review

**Comment by akshaydeo** [comment]
@coderabbitai full review

File: core/providers/bedrock/bedrock_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/bedrock/bedrock_test.go:19]: This is okay for now — these helpers only marshal static test fixtures so the error path is unreachable in practice. We'll align with the `must*` panic convention in a follow-up.

**Comment by akshaydeo** [comment]
### Merge activity

* **Mar 20, 4:42 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2174).
* **Mar 20, 4:43 PM UTC**: @akshaydeo merged this pull request with [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2174).
